### ISSUE TITLE
Type mismatch in fft_r2c

### ISSUE THREAD

**Issue by hayatoikoma** [gaika/ArrayFire.jl#3]
I just found that there is an issue for `fft_r2c`.

```
julia> a = rand(Float32, 10);

julia> ad = AFArray(a);

julia> bd = fft_r2c(ad, 1.0, 0)
ERROR: AssertionError: type mismatch: Complex{Float32} != Float32
 in ArrayFire.AFArray{Float32,1}(::Ptr{Void}) at /home/hayato/Development/JuliaPackage/tmp/ArrayFire.jl/src/array.jl:4
 in fft_r2c(::ArrayFire.AFArray{Float32,1}, ::Float64, ::Int64) at /home/hayato/Development/JuliaPackage/tmp/ArrayFire.jl/src/wrap.jl:1898
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
```

I have also tried
```
julia> complex_ad = AFArray(complex(a));

julia> bd = fft_r2c(complex_ad, 1.0, 0)
ERROR: ArrayFire Error (204) : Function does not support this data type
In function af_err fft_r2c(void**, af_array, double, dim_t, const dim_t*) [with int rank = 1; af_array = void*; dim_t = long long int]
In file src/api/c/fft.cpp:183
Invalid type for argument 1

 in _error(::UInt32) at /home/hayato/Development/JuliaPackage/tmp/ArrayFire.jl/src/util.jl:19
 in fft_r2c(::ArrayFire.AFArray{Complex{Float32},1}, ::Float64, ::Int64) at /home/hayato/Development/JuliaPackage/tmp/ArrayFire.jl/src/wrap.jl:1897
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
```

Since this implementation is different from the original implementation, I wasn't able to figure out how I can fix it immediately.

**Comment by gaika** [comment]
The generator had to be updated so correct type is returned. You can patch it with using AFArray!() constructor instead of inferred types. Please let me know if other functions need updates too. 

Fixed in 16be44b38edbf87935a4bc753acca76eddebe169

**Comment by hayatoikoma** [comment]
I am not 100% sure about what you meant.

Even if the inferred type in `wrap.jl` is wrong, a returned type is corrected based on the type returned by ArrayFire's C function? For `fft2_c2r`, the type of the input can be `ArrayFire{Float64,2}`, and the type of the output is inferred as `ArrayFire{Float64,2}`, which is wrong because it is a transform from a real array to a complex array. But the actual returned type was `ArrayFire{Complex{Float64}, 2}`, which looks correct.

I am also not sure what kind of behavior is expected when we supply a 2D array for one dimensional FFT related functions as an input. It returned something but couldn't figure out how it was transformed. But this question may be a question for the original ArrayFire....

**Comment by gaika** [comment]
Ok, c2r types are now fixed in c7e763ec95785f0170a28db1fdc8d84ffb223f81 

Ideally all ArrayFire.jl functions should be drop-in replacements for equivalent Julia functions. In this case it means the extra arguments should have default values that produce the same results in Julia.

Do you know what the mapping is for FFT?

**Comment by hayatoikoma** [comment]
Thank you. I didn't know `wrap.jl` was automatically generated.

I have a good understanding on their interface but don't know how they are implemented. At least, I know Julia's FFT interface and ArrayFire's FFT interface are different. For example, you can specify the direction of DFT for multi-dimensional array in Julia but cannot in ArrayFire. On the other hand, you can zero-pad an array to have a specific output size in ArrayFire but cannot in Julia.

It would be great to have the transform direction option in ArrayFire's FFT but I am not sure what is the most efficient way to do so. Anyway, it would be fantastic to use ArrayFire implementation without changing any Julia code.
